File: Java/src/sample/Item.java
package sample;

import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Item {
    private final SimpleStringProperty itemName;
    private final Set<Item> greater = new HashSet<>();

    public Item() {
        this(null);
    }

    public Item(String name) {
        this.itemName = new SimpleStringProperty(name);
    }

    public void setItemName(String itemName) {
        itemNameProperty().set(itemName);
    }

    public void addGreater(Item item) {
        List<Item> added = new ArrayList<>();
        added.add(item);
        item.greater.forEach( i -> {
            if(greater.add(i))
                added.add(i);
        });
        greater.add(item);
        Log.addLog(this, item, added);
    }
    public Set<Item> getGreater(){
        return greater;
    }
    public int getGreaterSize(){
        return greater.size();
    }

    public String getItemName() {
        return itemNameProperty().get();
    }

    public StringProperty itemNameProperty() {
        return itemName;
    }

    @Override
    public boolean equals(Object obj) {
        if (getItemName() == null && ((Item) obj).getItemName() == null)
            return true;
        if (obj == this) return true;
        if (!(obj instanceof Item))
            return false;
        Item item = (Item) obj;
        return getItemName().equals(((Item) obj).getItemName());
    }

    @Override
    public int hashCode() {
        return itemName.hashCode();
    }
}

File: js/Item.js
class Item {
    constructor(name) {
        this.greater = [];
        this.name = name;
    }

    setName(name) {
        this.name = name
    }

    getName() {
        return this.name;
    }

    addGreater(item) {
        let added = [];
        added.push(item);
        let exists = false;
        console.log(item);
        item.greater.forEach(element => {
            console.log(element);
            exists = false;
            this.greater.forEach(i => {
                if (element.getName() === i.getName()) {
                    exists = true;
                }
            });
            if (!exists) {
                this.greater.push(element);
                added.push(element);
            }
        });
        this.greater.push(item);
        Logs.addLog(this, item, added);
    }
}